overlord/exportstate/special.go

```diff
@@ -0,0 +1,172 @@
+		"snap-exec",           // used by snap-confine inside the per-snap mount namespace
+		"snap-gdb-shim",       // used by snap run --gdb
+		"snap-gdbserver-shim", // used by snap run --experimental-gdbserver
+		"snap-update-ns",      // used by snap-confine inside the per-snap mount namespace
+		"snapctl",             // used by snaps to talk to snapd
+	}
+	set := ExportSet{
+		Name:           "tools",
+		ConsumerIsHost: false, // Those tools are for other snaps to use.
+		Exports:        make(map[string]ExportedFile, len(tools)),
+	}
+	for _, tool := range tools {
+		set.Exports[tool] = ExportedFile{
+			Name:       tool,
+			SourcePath: filepath.Join(basePath, tool),
+		}
+	}
+	return set
+}
+
+// manifestForCoreSystem returns the manifest of the host as seen on ubuntu-core systems.
+func manifestForCoreSystem() *Manifest {
+	return &Manifest{
+		ExportedForSnapdAsVersion: "host", // Exception from the rule
+		SourceIsHost:              true,
+		// There are no export sets here, snapd snap is going to provide the tools.
+	}
+}
+
+// manifestForClassicSystem returns the manifest of the host as seen on classic systems.
+func manifestForClassicSystem() *Manifest {
+	tools := exportSetSnapdTools(dirs.DistroLibExecDir)
+	return &Manifest{
+		ExportedForSnapdAsVersion: "host", // Exception from the rule
+		SourceIsHost:              true,
+		Sets:                      map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForSnapdSnap returns the manifest of the snapd snap.
+func manifestForSnapdSnap(info *snap.Info) *Manifest {
+	tools := exportSetSnapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName: info.InstanceName(),
+		SnapRevision:     info.Revision,
+		Sets:             map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// manifestForCoreSnap returns the manifest of the core snap.
+func manifestForCoreSnap(info *snap.Info) *Manifest {
+	tools := exportSetSnapdTools("/usr/lib/snapd")
+	return &Manifest{
+		SnapInstanceName:          info.InstanceName(),
+		SnapRevision:              info.Revision,
+		ExportedForSnapdAsVersion: fmt.Sprintf("core_%s", info.Revision), // Exception from the rule
+
+		// Separate to avoid gofmt annoyance across versions.
+		Sets: map[string]ExportSet{tools.Name: tools},
+	}
+}
+
+// effectiveExportedVersionForSnapdOrCore returns the version to use for snapd tools export set or an error.
+//
+// The snapd tools export set is special as there are providers from snaps other
+// than snapd that need consideration. The result is, in order of preference:
+//
+// 0) "host" if on classic with disabled re-execution.
+// 1) snapd version, if available
+// 2) core version, if available
+// 3) "host" version, if on classic
+//
+// If no provider is available and there is nothing to export the returned
+// version is empty.
+func effectiveExportedVersionForSnapdOrCore(st *state.State) (exportedVersion string, err error) {
+	if release.OnClassic && os.Getenv("SNAP_REEXEC") == "0" {
```

Comment:
there was quite a bit of discussion around this in the original PR

Reply:
I'm referrring to this:

https://github.com/snapcore/snapd/pull/9546/files#r525943883

Reply:
Thanks, I see - I moved this to use snapdtool.IsReexec() now but given that having this here might be premature I will probably split it into the next PR instead.